code/modules/reagents/Chemistry-Holder.dm

```diff
@@ -407,9 +408,21 @@ var/global/obj/temp_reagents_holder = new
 	if(!amount)
 		return
 
-	var/part = amount / total_volume
+	var/part = amount
+	if(skip_reagents)
+		var/using_volume = total_volume
+		for(var/rtype in skip_reagents)
+			using_volume -= LAZYACCESS(reagent_volumes, rtype)
+		if(using_volume <= 0)
+			return
+		part /= using_volume
+	else
+		part /= total_volume
+
 	. = 0
 	for(var/rtype in reagent_volumes)
+		if(LAZYISIN(skip_reagents, rtype))
```

Comment:
this is O(mn), i wonder if it might be better to remove all entries in skip_reagents from a copy of reagent_volumes first? that would mean it'd be O(n+m), because we loop over m elements of skip_reagents followed by somewhere between n and n-m elements of reagent_volumes

actually judging from some quick wrench tests, you can just do `for(var/rtype in reagent_volumes - skip_reagents)`